src/helpers/puppets.ts

```diff
@@ -223,59 +235,80 @@ const getSplashScreenMetaData = async (
   options: Options,
 ): Promise<SavedImage[]> => {
+  let address: string;
+  let shellHtml: string;
+
   const logger = preLogger(saveImages.name, options);
   logger.log('Initialising puppeteer to take screenshots', '🤖');
 
-  const address = await url.getAddress(source, options);
-
-  return Promise.all(
-    imageList.map(async ({ name, width, height, scaleFactor, orientation }) => {
-      const browser = await puppeteer.launch({
-        headless: true,
-        defaultViewport: {
-          width,
-          height,
-        },
-        timeout: constants.BROWSER_SHELL_TIMEOUT,
-        args: constants.PUPPETEER_LAUNCH_ARGS,
-      });
-
-      const { type, quality } = options;
-
-      const path = file.getImageSavePath(name, output, type);
-
-      try {
-        const page = await browser.newPage();
-        await page.goto(address);
-        await page.screenshot({
-          path,
-          omitBackground: !options.opaque,
-          type: options.type,
-          ...(type !== 'png' ? { quality } : {}),
+  if (canNavigateTo(source)) {
+    address = await url.getAddress(source, options);
+  } else {
+    shellHtml = await url.getShellHtml(source, options);
+  }
+
+  return (
+    imageList
+      .map(async ({ name, width, height, scaleFactor, orientation }) => {
+        const { browser, chrome } = await browserHelper.getBrowserInstance({
+          defaultViewport: {
+            width,
+            height,
+          },
+          timeout: constants.BROWSER_SHELL_TIMEOUT,
         });
 
-        logger.success(`Saved image ${name}`);
-        return { name, width, height, scaleFactor, path, orientation };
-      } catch (e) {
-        logger.error(e.message);
-        throw Error(`Failed to save image ${name}`);
-      } finally {
-        await browser.close();
-      }
-    }),
+        const { type, quality } = options;
+        const path = file.getImageSavePath(name, output, type);
+
+        try {
+          const page = await browser.newPage();
+
+          if (address) {
+            await page.goto(address);
+          } else {
+            await page.setContent(shellHtml);
+          }
+
+          await page.screenshot({
+            path,
+            omitBackground: !options.opaque,
+            type: options.type,
+            ...(type !== 'png' ? { quality } : {}),
+          });
+
+          logger.success(`Saved image ${name}`);
+          await killBrowser(browser, chrome);
+
+          return { name, width, height, scaleFactor, path, orientation };
+        } catch (e) {
+          await killBrowser(browser, chrome);
+          logger.error(e.message);
+          throw Error(`Failed to save image ${name}`);
+        }
+      })
+      // Resolving array of promises in sequential manner to kill chrome instances properly
+      .reduce(
+        (acc, promise: Promise<SavedImage>) =>
+          acc.then(result => promise.then(Array.prototype.concat.bind(result))),
+        Promise.resolve([] as SavedImage[]),
```

Comment:
Hardcore `Promise.all` :)

Reply:
More like promise.map or promise.each bro, instead of running all instances in parallel, it executes and kills one by one.